src/EditorFeatures/Core/Implementation/IntelliSense/EditorCompletionImplementation/CompletionItemSource.cs

```diff
@@ -0,0 +1,249 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.Completion;
+using Microsoft.CodeAnalysis.Completion.Providers;
+using Microsoft.CodeAnalysis.Editor.Shared.Extensions;
+using Microsoft.CodeAnalysis.PooledObjects;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+using Microsoft.CodeAnalysis.Text;
+using Microsoft.CodeAnalysis.Text.Shared.Extensions;
+using Microsoft.VisualStudio.Core.Imaging;
+using Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion;
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Adornments;
+using Roslyn.Utilities;
+using EditorCompletion = Microsoft.VisualStudio.Language.Intellisense.AsyncCompletion.Data;
+using RoslynCompletionItem = Microsoft.CodeAnalysis.Completion.CompletionItem;
+using RoslynTrigger = Microsoft.CodeAnalysis.Completion.CompletionTrigger;
+
+namespace Microsoft.CodeAnalysis.Editor.Implementation.IntelliSense.Completion.EditorImplementation
+{
+    internal class CompletionItemSource : IAsyncCompletionSource
+    {
+        internal const string RoslynItem = nameof(RoslynItem);
+        internal const string TriggerBuffer = nameof(TriggerBuffer);
+        internal const string InsertionText = nameof(InsertionText);
+        internal const string MustSetSelection = nameof(MustSetSelection);
+        internal const string Description = nameof(Description);
+
+        private static ImmutableArray<ImageElement> s_WarningImageAttributeImagesArray;
+
+        public EditorCompletion.CompletionStartData InitializeCompletion(EditorCompletion.CompletionTrigger trigger, SnapshotPoint triggerLocation, CancellationToken cancellationToken)
+        {
+            var document = triggerLocation.Snapshot.GetOpenDocumentInCurrentContextWithChanges();
+            if (document == null)
+            {
+                return EditorCompletion.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            var service = document.GetLanguageService<CompletionService>();
+            if (service == null)
+            {
+                return EditorCompletion.CompletionStartData.DoesNotParticipateInCompletion;
+            }
+
+            var sourceText = document.GetTextSynchronously(cancellationToken);
+
+            // TODO: Check CompletionOptions.TriggerOnTyping  https://github.com/dotnet/roslyn/issues/27427
+            if (!(trigger.Reason == EditorCompletion.CompletionTriggerReason.Invoke ||
+                trigger.Reason == EditorCompletion.CompletionTriggerReason.InvokeAndCommitIfUnique)
+                && !service.ShouldTriggerCompletion(sourceText, triggerLocation.Position, GetRoslynTrigger(trigger, triggerLocation)))
```

Comment:
this expression is too complex. i can't tell what it's trying to actually test.  why not a more simple:

`if (reason != invoke && reason != invokeandcommit && !service.shoudltrigger) { return }` ?